src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -0,0 +1,251 @@
+            if (requestHandle.IsInvalid)
+            {
+                throw new HttpRequestException(SR.net_http_client_execution_error);
+            }
+
+            Interop.libcurl.curl_easy_setopt(requestHandle, CURLoption.CURLOPT_URL, state.RequestMessage.RequestUri.OriginalString);
+            if (_automaticRedirection)
+            {
+                Interop.libcurl.curl_easy_setopt(requestHandle, CURLoption.CURLOPT_FOLLOWLOCATION, 1);
+            }
+
+            // TODO: Handle headers, proxy and other options
+
+            return requestHandle;
+        }
+
+
+        private HttpResponseMessage CreateResponseMessage(SafeCurlHandle requestHandle, HttpRequestMessage request)
+        {
+            var response = new HttpResponseMessage();
+            Int64 httpStatusCode = 0;
+            int result = Interop.libcurl.curl_easy_getinfo(requestHandle, CURLINFO.CURLINFO_RESPONSE_CODE, ref httpStatusCode);
+            if (CURLcode.CURLE_OK != result)
+            {
+                throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result));
+            }
+            response.StatusCode = (HttpStatusCode)httpStatusCode;
+
+            // TODO: Do error processing if needed and return actual response
+            response.Content = new StringContent("SendAsync to " + request.RequestUri.OriginalString + " returned: " + response.StatusCode);
+            return response;
+        }
+
+        private void HandleAsyncException(RequestState state, Exception ex)
+        {
+            if (state.CancellationToken.IsCancellationRequested)
+            {
+                // If the exception was due to the cancellation token being canceled, throw cancellation exception.
+                state.TrySetCanceled();
+            }
+            else if (ex is HttpRequestException)
+            {
+                state.TrySetException(ex);
+            }
+            else
+            {
+                state.TrySetException(new HttpRequestException(SR.net_http_client_execution_error, ex));
+            }
+        }
+
+        private void CheckDisposed()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(GetType().FullName);
+            }
+        }
+
+        private void CheckDisposedOrStarted()
+        {
+            CheckDisposed();
+            if (_anyOperationStarted)
+            {
+                throw new InvalidOperationException(SR.net_http_operation_started);
+            }
+        }
+
+        private string GetCurlErrorString(int code)
+        {
+            IntPtr ptr = Interop.libcurl.curl_easy_strerror(code);
+            return Marshal.PtrToStringAnsi(ptr);
+        }
+
+        private Exception GetCurlException(int code)
+        {
+            return new Exception(GetCurlErrorString(code));
+        }
+
+        #endregion
+
+        private class RequestState : TaskCompletionSource<HttpResponseMessage>
```

Comment:
Minor since this is all internal/private, but wondering if `RequestState` should be renamed to something else with `CompletionSource` as a suffix, to make it more clear where it's used elsewhere that this is a subclass of `TaskCompletionSource<TResult>`.

A few ideas:
 - `RequestCompletionSource` (given the existing `RequestState` name)
 - `ResponseCompletionSource` (given that `TResult` is `HttpResponseMessage`)
 - `CurlCompletionSource`

This would be consistent with how a similar subclass, [`FileStreamCompletionSource`](https://github.com/dotnet/corefx/blob/e66911b80621bf2b53f7b33e170ad18b3741426c/src/System.IO.FileSystem/src/System/IO/Win32FileStreamCompletionSource.cs#L18), is named.